src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -296,5 +284,139 @@ internal SafeSslHandle(IntPtr validSslPointer, bool ownsHandle) : base(IntPtr.Ze
         {
             handle = validSslPointer;
         }
+
+        internal class ReadBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private int _bytesAvailable;
+            private byte[] _byteArray;
+            private int _offset;
+
+            internal ReadBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.CustomBio.BioSetGCHandle(_bioHandle, _handle);
+                Interop.Crypto.BioSetFlags(bioHandle, Interop.Crypto.BIO_FLAGS.BIO_FLAGS_READ | Interop.Crypto.BIO_FLAGS.BIO_FLAGS_SHOULD_RETRY);
+            }
+
+            public void SetBio(byte[] buffer, int offset, int length)
+            {
+                _byteArray = buffer;
+                _offset = offset;
+                _bytesAvailable = length;
+            }
+
+            public int Read(Span<byte> output)
+            {
+                var bytesToCopy = Math.Min(output.Length, _bytesAvailable);
+                if (bytesToCopy == 0) return -1;
+                var span = new Span<byte>(_byteArray, _offset, bytesToCopy);
+                span.CopyTo(output);
+                _offset += bytesToCopy;
+                _bytesAvailable -= bytesToCopy;
+                return bytesToCopy;
+            }
+
+            //Bio is already released by the ssl object
+            private void Dispose(bool isDisposing)
+            {
+                if (_handle.IsAllocated)
+                {
+                    _handle.Free();
+                }
+            }
+
+            public void Dispose()
+            {
+                Dispose(false);
+                GC.SuppressFinalize(this);
+            }
+
+            ~ReadBioBuffer()
+            {
+                Dispose(false);
+            }
+        }
+
+        internal class WriteBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private byte[] _byteArray;
+            private int _bytesWritten;
+            private bool _isHandshake;
+
+            internal WriteBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.CustomBio.BioSetGCHandle(_bioHandle, _handle);
+            }
+
+            public void SetBio(byte[] buffer, bool isHandshake)
+            {
+                _byteArray = buffer;
+                _bytesWritten = 0;
+                _isHandshake = isHandshake;
+            }
+
+            public int TakeBytes(out byte[] output)
+            {
+                var bytes = _bytesWritten;
```

Comment:
Race conditions.  Like Write sees `_byteArray != null` right after TakeBytes set `_bytesWritten` to 0, but before `_byteArray = null`.  Since neither of these methods works on/with pointers directly it might not be necessary since the class is (IIRC) described as not thread-safe.